{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "02a74dff",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import glob\n",
    "import numpy as np\n",
    "import statistics as stats\n",
    "import scipy\n",
    "import mpra_tools.predicted_occupancy as po\n",
    "import mpra_tools.fasta_utils as fu\n",
    "import math\n",
    "import matplotlib.pyplot as plt\n",
    "from collections import Counter, defaultdict\n",
    "import random\n",
    "from sklearn.metrics import precision_recall_curve, roc_curve, f1_score, precision_score, recall_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8624e4e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "118364 samples\n"
     ]
    }
   ],
   "source": [
    "activity_df = pd.read_csv(\"Data/activity.csv\", index_col=0)\n",
    "retinopathy_df = pd.read_csv(\"Data/retinopathy_reformatted.txt\", sep='\\t', index_col=1)\n",
    "test_labels = activity_df[activity_df['test_set'] | activity_df['cnn_validation_set']]['label']\n",
    "train_labels = activity_df[~activity_df['label'].isin(test_labels)]['label']\n",
    "L = 164\n",
    "print(len(activity_df), \"samples\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0f02db19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>library</th>\n",
       "      <th>genotype</th>\n",
       "      <th>activity_mean</th>\n",
       "      <th>activity_std</th>\n",
       "      <th>n_observations</th>\n",
       "      <th>activity_mu</th>\n",
       "      <th>activity_sigma</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>activity_class</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>chip_peak_id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>activity_vs_basal</th>\n",
       "      <th>expression_log2</th>\n",
       "      <th>activity_bin</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CRX.1003</th>\n",
       "      <td>0</td>\n",
       "      <td>rho</td>\n",
       "      <td>WT</td>\n",
       "      <td>0.212773</td>\n",
       "      <td>0.069493</td>\n",
       "      <td>12.0</td>\n",
       "      <td>-1.598208</td>\n",
       "      <td>0.318364</td>\n",
       "      <td>4.222015e-02</td>\n",
       "      <td>5.154038e-02</td>\n",
       "      <td>inactive</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>CRX.1003</td>\n",
       "      <td>CCAGCGTTCCTCCCATCTGTAGCAGTCAGGGTTCTTAAGTAGCAAA...</td>\n",
       "      <td>0.771869</td>\n",
       "      <td>-0.373573</td>\n",
       "      <td>Inactive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CRX.1004</th>\n",
       "      <td>2</td>\n",
       "      <td>rho</td>\n",
       "      <td>WT</td>\n",
       "      <td>1.056873</td>\n",
       "      <td>0.300882</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.016349</td>\n",
       "      <td>0.279162</td>\n",
       "      <td>8.650884e-13</td>\n",
       "      <td>2.948816e-12</td>\n",
       "      <td>weak_enhancer</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>CRX.1004</td>\n",
       "      <td>CTAACTTCCGTATCCACTGCCCCCACCCCCACCCCTGTGTCCTCAC...</td>\n",
       "      <td>3.833980</td>\n",
       "      <td>1.938843</td>\n",
       "      <td>WeakEnhancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CRX.1007</th>\n",
       "      <td>4</td>\n",
       "      <td>rho</td>\n",
       "      <td>WT</td>\n",
       "      <td>0.010203</td>\n",
       "      <td>0.010766</td>\n",
       "      <td>12.0</td>\n",
       "      <td>-4.959274</td>\n",
       "      <td>0.865048</td>\n",
       "      <td>7.189995e-09</td>\n",
       "      <td>1.738871e-08</td>\n",
       "      <td>strong_silencer</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>CRX.1007</td>\n",
       "      <td>CCTGCTAACCTGACCTTCTTATTCATAGATGAGGAAATAGAGGCTT...</td>\n",
       "      <td>0.037011</td>\n",
       "      <td>-4.755889</td>\n",
       "      <td>Silencer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CRX.101</th>\n",
       "      <td>6</td>\n",
       "      <td>rho</td>\n",
       "      <td>WT</td>\n",
       "      <td>0.910381</td>\n",
       "      <td>0.369439</td>\n",
       "      <td>12.0</td>\n",
       "      <td>-0.170115</td>\n",
       "      <td>0.390443</td>\n",
       "      <td>3.512835e-08</td>\n",
       "      <td>7.924102e-08</td>\n",
       "      <td>weak_enhancer</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>CRX.101</td>\n",
       "      <td>CACCGCCCACAGGACACCAGCCCTTGGGTTTTATAATCTCTCACAG...</td>\n",
       "      <td>3.302558</td>\n",
       "      <td>1.723584</td>\n",
       "      <td>WeakEnhancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CRX.1011</th>\n",
       "      <td>7</td>\n",
       "      <td>rho</td>\n",
       "      <td>WT</td>\n",
       "      <td>0.663029</td>\n",
       "      <td>0.559840</td>\n",
       "      <td>12.0</td>\n",
       "      <td>-0.680046</td>\n",
       "      <td>0.733635</td>\n",
       "      <td>8.350663e-03</td>\n",
       "      <td>1.092875e-02</td>\n",
       "      <td>weak_enhancer</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>CRX.1011</td>\n",
       "      <td>GGCAGGCTTTGTTGTCGTCCGAGCAGCTGTGACCGTTGGGGCTGAG...</td>\n",
       "      <td>2.405247</td>\n",
       "      <td>1.266185</td>\n",
       "      <td>WeakEnhancer</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Unnamed: 0 library genotype  activity_mean  activity_std  \\\n",
       "label                                                                \n",
       "CRX.1003           0     rho       WT       0.212773      0.069493   \n",
       "CRX.1004           2     rho       WT       1.056873      0.300882   \n",
       "CRX.1007           4     rho       WT       0.010203      0.010766   \n",
       "CRX.101            6     rho       WT       0.910381      0.369439   \n",
       "CRX.1011           7     rho       WT       0.663029      0.559840   \n",
       "\n",
       "          n_observations  activity_mu  activity_sigma        pvalue  \\\n",
       "label                                                                 \n",
       "CRX.1003            12.0    -1.598208        0.318364  4.222015e-02   \n",
       "CRX.1004            12.0     0.016349        0.279162  8.650884e-13   \n",
       "CRX.1007            12.0    -4.959274        0.865048  7.189995e-09   \n",
       "CRX.101             12.0    -0.170115        0.390443  3.512835e-08   \n",
       "CRX.1011            12.0    -0.680046        0.733635  8.350663e-03   \n",
       "\n",
       "                qvalue   activity_class variant_type chip_peak_id  \\\n",
       "label                                                               \n",
       "CRX.1003  5.154038e-02         inactive     wildtype     CRX.1003   \n",
       "CRX.1004  2.948816e-12    weak_enhancer     wildtype     CRX.1004   \n",
       "CRX.1007  1.738871e-08  strong_silencer     wildtype     CRX.1007   \n",
       "CRX.101   7.924102e-08    weak_enhancer     wildtype      CRX.101   \n",
       "CRX.1011  1.092875e-02    weak_enhancer     wildtype     CRX.1011   \n",
       "\n",
       "                                                   sequence  \\\n",
       "label                                                         \n",
       "CRX.1003  CCAGCGTTCCTCCCATCTGTAGCAGTCAGGGTTCTTAAGTAGCAAA...   \n",
       "CRX.1004  CTAACTTCCGTATCCACTGCCCCCACCCCCACCCCTGTGTCCTCAC...   \n",
       "CRX.1007  CCTGCTAACCTGACCTTCTTATTCATAGATGAGGAAATAGAGGCTT...   \n",
       "CRX.101   CACCGCCCACAGGACACCAGCCCTTGGGTTTTATAATCTCTCACAG...   \n",
       "CRX.1011  GGCAGGCTTTGTTGTCGTCCGAGCAGCTGTGACCGTTGGGGCTGAG...   \n",
       "\n",
       "          activity_vs_basal  expression_log2  activity_bin  \n",
       "label                                                       \n",
       "CRX.1003           0.771869        -0.373573      Inactive  \n",
       "CRX.1004           3.833980         1.938843  WeakEnhancer  \n",
       "CRX.1007           0.037011        -4.755889      Silencer  \n",
       "CRX.101            3.302558         1.723584  WeakEnhancer  \n",
       "CRX.1011           2.405247         1.266185  WeakEnhancer  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "retinopathy_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "869dfed9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "273886 motifs\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['NDF1', 'CRX', 'RORB', 'GFI1', 'MAZ', 'NRL', 'MEF2D', 'RAX']"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fimo_df = pd.concat([pd.read_csv('Data/fimo_eLife_activity/fimo.tsv', sep='\\t'),pd.read_csv('Data/fimo_eLife_retinopathy/fimo.tsv', sep='\\t')], ignore_index=True)\n",
    "del fimo_df['motif_alt_id']\n",
    "fimo_df.dropna(inplace=True)\n",
    "fimo_df['motif_id'] = fimo_df['motif_id'].map(lambda x: x.split('_')[0])\n",
    "fimo_df = fimo_df.astype({'start':int, 'stop':int})\n",
    "print(len(fimo_df),\"motifs\")\n",
    "list(set(fimo_df['motif_id']))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "155eab30",
   "metadata": {},
   "source": [
    "### Transform Each Sequence into a sentence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "50103e95",
   "metadata": {},
   "outputs": [],
   "source": [
    "# main dataset\n",
    "labels = fimo_df.groupby(by=\"sequence_name\")\n",
    "sentences = dict()\n",
    "IVAL = 5.0\n",
    "\n",
    "for name, df in labels:\n",
    "    sdf = df.sort_values(by=\"start\")\n",
    "    i = 1\n",
    "    s = []\n",
    "    for index, row in sdf.iterrows():\n",
    "        d = row['start']-i\n",
    "        if d > 0:\n",
    "            xIVALmer = math.ceil(d/IVAL)\n",
    "            s.append(xIVALmer)\n",
    "        s.append(row['motif_id']+row['strand'])\n",
    "        i = row['stop']+1\n",
    "    if i < L:\n",
    "        xIVALmer = math.ceil((L-i)/IVAL)\n",
    "        s.append(xIVALmer)\n",
    "    sentences[name] = s\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7cb9193",
   "metadata": {},
   "source": [
    "### Get word counts for each document class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "61f37718",
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = ['Silencer','Inactive','WeakEnhancer','StrongEnhancer']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "30a5e2a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Total word counts for each class\n",
    "counts = dict(\n",
    "    [(b,Counter()) for b in bins]\n",
    ")\n",
    "\n",
    "\n",
    "#Proportion of word counts N(w)/N(V)\n",
    "p_wc = dict(\n",
    "    [(b,defaultdict(float)) for b in bins]\n",
    ")\n",
    "\n",
    "\n",
    "#Counts all words for each class\n",
    "train_labels = train_labels[train_labels.isin(sentences.keys())]\n",
    "train_bins = activity_df.loc[train_labels.index]['activity_bin']\n",
    "\n",
    "for i in train_labels.index:\n",
    "    s = sentences[train_labels.loc[i]]\n",
    "    counts[train_bins.loc[i]].update(s)\n",
    "\n",
    "\n",
    "#Convert raw counts to naive probabilities\n",
    "for b in bins:\n",
    "    p_wc[b] = {k: v / counts[b].total() for k,v in counts[b].items()}\n",
    "    \n",
    "    \n",
    "#Get class probabilities\n",
    "bin_counts = activity_df.loc[train_labels.index]['activity_bin'].value_counts()\n",
    "p_c = dict(\n",
    "    [(b,bin_counts[b]/sum(bin_counts)) for b in bins]\n",
    ")\n",
    "\n",
    "\n",
    "#Get length of alphabet\n",
    "alphabet = list(set(np.concatenate(list(sentences.values()))))\n",
    "V = len(alphabet)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97e674aa",
   "metadata": {},
   "source": [
    "### Get class probabilities for all docs in the testing set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "7c72d5be",
   "metadata": {},
   "outputs": [],
   "source": [
    "preds_test = dict()\n",
    "truths_test = dict()\n",
    "\n",
    "test_labels = test_labels[test_labels.isin(sentences.keys())]\n",
    "test_bins = activity_df.loc[test_labels.index]['activity_bin']\n",
    "\n",
    "for i in test_labels.index:\n",
    "    #Inititial probabilities for each doc P(c_i)\n",
    "    probs = dict([(b,p_c[b]) for b in bins])\n",
    "    for b in bins:\n",
    "        for w in sentences[test_labels.loc[i]]:\n",
    "            # Probability that a word appears in the doc.  Log transform means we can add\n",
    "            probs[b] = probs[b] * (counts[b][w]+1)/(counts[b].total()+V)\n",
    "    norm = sum(probs.values())\n",
    "    preds_test[test_labels.loc[i]] = np.array(list(probs.values())) / norm\n",
    "    truths_test[test_labels.loc[i]] = bins.index(test_bins.loc[i])\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "f00a6969",
   "metadata": {},
   "outputs": [],
   "source": [
    "preds_ret = dict()\n",
    "truths_ret = dict()\n",
    "\n",
    "ret_labels = retinopathy_df[retinopathy_df.index.isin(sentences.keys())].index\n",
    "ret_bins = retinopathy_df.loc[ret_labels]['activity_bin']\n",
    "\n",
    "for i in ret_labels:\n",
    "    #Inititial probabilities for each doc P(c_i)\n",
    "    probs = dict([(b,p_c[b]) for b in bins])\n",
    "    for b in bins:\n",
    "        for w in sentences[i]:\n",
    "            # Probability that a word appears in the doc.  Log transform means we can add\n",
    "            probs[b] = probs[b] * (counts[b][w]+1)/(counts[b].total()+V)\n",
    "    norm = sum(probs.values())\n",
    "    preds_ret[i] = np.array(list(probs.values())) / norm\n",
    "    truths_ret[i] = bins.index(ret_bins.loc[i])\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "db358a7c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "retinopathy test\n",
      "correct: 0.32\n"
     ]
    }
   ],
   "source": [
    "averages = ['micro', 'macro','weighted']\n",
    "\n",
    "print(\"retinopathy test\")\n",
    "t = list(truths_ret.values())\n",
    "p = [a.argmax() for a in preds_ret.values()]\n",
    "\n",
    "correct = 0\n",
    "for truth, pred in zip(t,p):\n",
    "    if truth == pred:\n",
    "        correct += 1\n",
    "        \n",
    "correct = correct/len(t)\n",
    "print(\"correct:\", round(correct,2))\n",
    "\n",
    "\n",
    "f1_ret = [f1_score(t,p, labels=range(len(bins)), average=a) for a in averages] + [correct]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "d2643b66",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "t = list(truths_test.values())\n",
    "p = [a.argmax() for a in preds_test.values()]\n",
    "\n",
    "correct = 0\n",
    "for truth, pred in zip(t,p):\n",
    "    if truth == pred:\n",
    "        correct += 1\n",
    "        \n",
    "correct = correct/len(t)\n",
    "\n",
    "f1_test = [f1_score(t,p, labels=range(len(bins)), average=a) for a in averages] + [correct]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "f3b5822d",
   "metadata": {},
   "outputs": [],
   "source": [
    "bayes_preformance = pd.DataFrame(\n",
    "    data = [f1_ret,f1_test],\n",
    "    columns=averages+['correct'],\n",
    "    index=['Retinopathy Set', 'Test Set']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "b72201b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "Clf_performance = pd.read_csv(\"Data/cnn_dataset_performance_metrics.txt\", sep ='\\t', index_col=[0,1])\n",
    "clf_test = Clf_performance[Clf_performance['test_set']=='test_set'].copy()\n",
    "clf_ret = Clf_performance[Clf_performance['test_set']=='test_retinopathy'].copy()\n",
    "clf_test.drop(columns=['test_set','nseqs_train','nseqs_test'],inplace=True)\n",
    "clf_ret.drop(columns=['test_set','nseqs_train','nseqs_test'],inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "d4c0c6c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>micro</th>\n",
       "      <th>macro</th>\n",
       "      <th>weighted</th>\n",
       "      <th>Si</th>\n",
       "      <th>In</th>\n",
       "      <th>WE</th>\n",
       "      <th>SE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dataset</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GenomicOnly</th>\n",
       "      <td>0.350914</td>\n",
       "      <td>0.145846</td>\n",
       "      <td>0.211773</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.506837</td>\n",
       "      <td>0.071107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Round1</th>\n",
       "      <td>0.362166</td>\n",
       "      <td>0.172657</td>\n",
       "      <td>0.241627</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.504582</td>\n",
       "      <td>0.143642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Round2</th>\n",
       "      <td>0.354430</td>\n",
       "      <td>0.135631</td>\n",
       "      <td>0.191908</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015566</td>\n",
       "      <td>0.525931</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Round3a</th>\n",
       "      <td>0.371308</td>\n",
       "      <td>0.268965</td>\n",
       "      <td>0.370702</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.271604</td>\n",
       "      <td>0.327779</td>\n",
       "      <td>0.467166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Round3aNoRound2</th>\n",
       "      <td>0.401547</td>\n",
       "      <td>0.282342</td>\n",
       "      <td>0.395492</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.231200</td>\n",
       "      <td>0.439486</td>\n",
       "      <td>0.456208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Round3b</th>\n",
       "      <td>0.440928</td>\n",
       "      <td>0.341651</td>\n",
       "      <td>0.446962</td>\n",
       "      <td>0.067816</td>\n",
       "      <td>0.355312</td>\n",
       "      <td>0.387801</td>\n",
       "      <td>0.563356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Round3c</th>\n",
       "      <td>0.341772</td>\n",
       "      <td>0.175502</td>\n",
       "      <td>0.215086</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.171492</td>\n",
       "      <td>0.476005</td>\n",
       "      <td>0.058765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Round4a</th>\n",
       "      <td>0.317159</td>\n",
       "      <td>0.174829</td>\n",
       "      <td>0.223549</td>\n",
       "      <td>0.066964</td>\n",
       "      <td>0.039871</td>\n",
       "      <td>0.518804</td>\n",
       "      <td>0.065261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Round4b</th>\n",
       "      <td>0.345992</td>\n",
       "      <td>0.282339</td>\n",
       "      <td>0.322559</td>\n",
       "      <td>0.144871</td>\n",
       "      <td>0.266884</td>\n",
       "      <td>0.477378</td>\n",
       "      <td>0.234446</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    micro     macro  weighted        Si        In        WE  \\\n",
       "dataset                                                                       \n",
       "GenomicOnly      0.350914  0.145846  0.211773  0.000000  0.000000  0.506837   \n",
       "Round1           0.362166  0.172657  0.241627  0.000000  0.000000  0.504582   \n",
       "Round2           0.354430  0.135631  0.191908  0.000000  0.015566  0.525931   \n",
       "Round3a          0.371308  0.268965  0.370702  0.000000  0.271604  0.327779   \n",
       "Round3aNoRound2  0.401547  0.282342  0.395492  0.000000  0.231200  0.439486   \n",
       "Round3b          0.440928  0.341651  0.446962  0.067816  0.355312  0.387801   \n",
       "Round3c          0.341772  0.175502  0.215086  0.000000  0.171492  0.476005   \n",
       "Round4a          0.317159  0.174829  0.223549  0.066964  0.039871  0.518804   \n",
       "Round4b          0.345992  0.282339  0.322559  0.144871  0.266884  0.477378   \n",
       "\n",
       "                       SE  \n",
       "dataset                    \n",
       "GenomicOnly      0.071107  \n",
       "Round1           0.143642  \n",
       "Round2           0.000000  \n",
       "Round3a          0.467166  \n",
       "Round3aNoRound2  0.456208  \n",
       "Round3b          0.563356  \n",
       "Round3c          0.058765  \n",
       "Round4a          0.065261  \n",
       "Round4b          0.234446  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf_test.groupby(level=0).agg('median')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "b5b56cb0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>micro</th>\n",
       "      <th>macro</th>\n",
       "      <th>weighted</th>\n",
       "      <th>Si</th>\n",
       "      <th>In</th>\n",
       "      <th>WE</th>\n",
       "      <th>SE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dataset</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GenomicOnly</th>\n",
       "      <td>0.483749</td>\n",
       "      <td>0.176835</td>\n",
       "      <td>0.335919</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004785</td>\n",
       "      <td>0.657935</td>\n",
       "      <td>0.033794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Round1</th>\n",
       "      <td>0.470981</td>\n",
       "      <td>0.193926</td>\n",
       "      <td>0.340705</td>\n",
       "      <td>0.016178</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.648187</td>\n",
       "      <td>0.109796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Round2</th>\n",
       "      <td>0.473012</td>\n",
       "      <td>0.184800</td>\n",
       "      <td>0.333342</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.093793</td>\n",
       "      <td>0.645405</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Round3a</th>\n",
       "      <td>0.183401</td>\n",
       "      <td>0.160619</td>\n",
       "      <td>0.155211</td>\n",
       "      <td>0.002062</td>\n",
       "      <td>0.206688</td>\n",
       "      <td>0.215517</td>\n",
       "      <td>0.214100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Round3aNoRound2</th>\n",
       "      <td>0.281776</td>\n",
       "      <td>0.219404</td>\n",
       "      <td>0.271737</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.215467</td>\n",
       "      <td>0.450504</td>\n",
       "      <td>0.222348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Round3b</th>\n",
       "      <td>0.232443</td>\n",
       "      <td>0.207965</td>\n",
       "      <td>0.219153</td>\n",
       "      <td>0.031886</td>\n",
       "      <td>0.220859</td>\n",
       "      <td>0.308574</td>\n",
       "      <td>0.254433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Round3c</th>\n",
       "      <td>0.432966</td>\n",
       "      <td>0.192154</td>\n",
       "      <td>0.328411</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.102592</td>\n",
       "      <td>0.622209</td>\n",
       "      <td>0.049745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Round4a</th>\n",
       "      <td>0.505514</td>\n",
       "      <td>0.314424</td>\n",
       "      <td>0.473670</td>\n",
       "      <td>0.522945</td>\n",
       "      <td>0.084478</td>\n",
       "      <td>0.626301</td>\n",
       "      <td>0.010363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Round4b</th>\n",
       "      <td>0.370284</td>\n",
       "      <td>0.338656</td>\n",
       "      <td>0.396018</td>\n",
       "      <td>0.495700</td>\n",
       "      <td>0.178315</td>\n",
       "      <td>0.421774</td>\n",
       "      <td>0.246826</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    micro     macro  weighted        Si        In        WE  \\\n",
       "dataset                                                                       \n",
       "GenomicOnly      0.483749  0.176835  0.335919  0.000000  0.004785  0.657935   \n",
       "Round1           0.470981  0.193926  0.340705  0.016178  0.000000  0.648187   \n",
       "Round2           0.473012  0.184800  0.333342  0.000000  0.093793  0.645405   \n",
       "Round3a          0.183401  0.160619  0.155211  0.002062  0.206688  0.215517   \n",
       "Round3aNoRound2  0.281776  0.219404  0.271737  0.000000  0.215467  0.450504   \n",
       "Round3b          0.232443  0.207965  0.219153  0.031886  0.220859  0.308574   \n",
       "Round3c          0.432966  0.192154  0.328411  0.000000  0.102592  0.622209   \n",
       "Round4a          0.505514  0.314424  0.473670  0.522945  0.084478  0.626301   \n",
       "Round4b          0.370284  0.338656  0.396018  0.495700  0.178315  0.421774   \n",
       "\n",
       "                       SE  \n",
       "dataset                    \n",
       "GenomicOnly      0.033794  \n",
       "Round1           0.109796  \n",
       "Round2           0.000000  \n",
       "Round3a          0.214100  \n",
       "Round3aNoRound2  0.222348  \n",
       "Round3b          0.254433  \n",
       "Round3c          0.049745  \n",
       "Round4a          0.010363  \n",
       "Round4b          0.246826  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf_ret.groupby(level=0).agg('median')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "78ebcc19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>micro</th>\n",
       "      <th>macro</th>\n",
       "      <th>weighted</th>\n",
       "      <th>correct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Retinopathy Set</th>\n",
       "      <td>0.318545</td>\n",
       "      <td>0.218040</td>\n",
       "      <td>0.258243</td>\n",
       "      <td>0.318545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Test Set</th>\n",
       "      <td>0.262172</td>\n",
       "      <td>0.219618</td>\n",
       "      <td>0.275600</td>\n",
       "      <td>0.262172</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    micro     macro  weighted   correct\n",
       "Retinopathy Set  0.318545  0.218040  0.258243  0.318545\n",
       "Test Set         0.262172  0.219618  0.275600  0.262172"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bayes_preformance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d46fd621",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3c677bf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dbc725e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54471130",
   "metadata": {},
   "outputs": [],
   "source": [
    "seqs = activity_df['sequence']\n",
    "\n",
    "motifs = 0\n",
    "\n",
    "for seq in seqs[:1000]:\n",
    "    motif_dict = po.get_occupied_sites_and_tfs(po.total_landscape(seq,ewm,9), cutoff=0.2)\n",
    "    motifs += len(motif_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5963d1dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.21273408239700375"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "correct = 0\n",
    "for truth in t:\n",
    "    pred = random.choices([0,1,2,3],weights = list(p_c.values()))[0]\n",
    "    if truth == pred:\n",
    "        correct += 1\n",
    "\n",
    "correct/len(t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "934cb8a3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e11d9718",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f635de4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89837231",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26794871",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "943c9e9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "ewm = po.read_pwm_to_ewm(\"Data/Motifs/eLifeMotifs.meme\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e1988ac5",
   "metadata": {},
   "outputs": [],
   "source": [
    "seqs = activity_df['sequence']\n",
    "\n",
    "motifs = 0\n",
    "\n",
    "for seq in seqs[:1000]:\n",
    "    motif_dict = po.get_occupied_sites_and_tfs(po.total_landscape(seq,ewm,9), cutoff=0.2)\n",
    "    motifs += len(motif_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c27692c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3069"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "301c964b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
